tests/test_frame.py

```diff
@@ -1761,7 +1761,8 @@ def test_getmultiindex_columns() -> None:
         [(i, s) for i in [1] for s in df.columns.get_level_values(1)]
     ]
     res4: pd.DataFrame = df[[df.columns[0]]]
-    check(assert_type(df[df.columns[0]], pd.Series), pd.Series)
+    column: Scalar = df.columns[0]
+    check(assert_type(df[column], pd.Series), pd.Series)
```

Comment:
I'm not sure if this can be fixed - maybe we can do some gymnastic with the order of overloads. `df.columns` is `Index[Any]` and would therefore probably match the first overload.

Reply:
Given that both mypy and pyright do not limit a bounded TypeVar to its bound when it is unknown/any, we have only two options:

- make people cast the Index to the appropriate type
- return Scalar
- return Scalar now; in the future, make pd.DataFrame generic in terms of the Index; return S1

I have no particular preference, except that making DataFrame generic might not happen anytime soon

btw. I will offline from Friday-Sunday, you are welcome to push changes to this PR or also merge it. Since this is a large PR and the tests might not cover everything, I would prefer if you or others could run this version of pandas-stubs on your internal codebase before the next release to catch potential regressions.

Reply:
unrelated: seems that the newest version of numexp broke CI/pandas

Reply:
> Given that both mypy and pyright do not limit a bounded TypeVar to its bound when it is unknown/any, we have only two options:
> 
> * make people cast the Index to the appropriate type
> * return Scalar
> * return Scalar now; in the future, make pd.DataFrame generic in terms of the Index; return S1
> 
> I have no particular preference, except that making DataFrame generic might not happen anytime soon

Would changing `DataFrame.columns()` to return `Index[Scalar]` work?

> 
> btw. I will offline from Friday-Sunday, you are welcome to push changes to this PR or also merge it. Since this is a large PR and the tests might not cover everything, I would prefer if you or others could run this version of pandas-stubs on your internal codebase before the next release to catch potential regressions.

I will see if I can give this a try.

Reply:
> Since this is a large PR and the tests might not cover everything, I would prefer if you or others could run this version of pandas-stubs on your internal codebase before the next release to catch potential regressions.

I tried the version I placed in your repo with the PR on two large code bases and no new errors appeared, so I think once you merge that in, I can approve and merge in this PR.

Reply:
Thank you for testing it and finding a way to let mypy (and pyright) clearly indicate the unintended calls!